src/app/tests/TestCommandInteraction.cpp

```diff
@@ -589,6 +611,119 @@ void TestCommandInteraction::TestCommandHandlerCommandEncodeFailure(nlTestSuite
+    commandSender.MoveToState(app::CommandSender::State::ResponseReceived);
+
+    System::PacketBufferHandle msgBuf = System::PacketBufferHandle::New(kMaxSecureSduLengthBytes);
+    NL_TEST_ASSERT(apSuite, !msgBuf.IsNull());
+    System::PacketBufferTLVWriter writer;
+    writer.Init(std::move(msgBuf));
+    StatusResponseMessage::Builder response;
+    response.Init(&writer);
+    response.Status(Protocols::InteractionModel::Status::InvalidAction);
+    NL_TEST_ASSERT(apSuite, writer.Finalize(&msgBuf) == CHIP_NO_ERROR);
+
+    PayloadHeader payloadHeader;
+    payloadHeader.SetExchangeID(0);
+    payloadHeader.SetMessageType(chip::Protocols::InteractionModel::MsgType::StatusResponse);
+
+    commandSender.OnMessageReceived(commandSender.mpExchangeCtx, payloadHeader, std::move(msgBuf));
+
+    NL_TEST_ASSERT(apSuite,
+                   mockCommandSenderDelegate.onResponseCalledTimes == 0 && mockCommandSenderDelegate.onFinalCalledTimes == 1 &&
+                       mockCommandSenderDelegate.onErrorCalledTimes == 1);
+
+    // Decrease CommandHandler refcount and send response
+    asyncCommandHandle = nullptr;
+
+    ctx.DrainAndServiceIO();
+
+    NL_TEST_ASSERT(apSuite, GetNumActiveHandlerObjects() == 0);
+    NL_TEST_ASSERT(apSuite, ctx.GetExchangeManager().GetNumActiveExchanges() == 0);
+}
+
+// Command Sender sends the invoke request, handler runs the async procedure, and triggers unknown message call
+void TestCommandInteraction::TestCommandInvalidMessage2(nlTestSuite * apSuite, void * apContext)
+{
+    TestContext & ctx = *static_cast<TestContext *>(apContext);
+    CHIP_ERROR err    = CHIP_NO_ERROR;
+
+    mockCommandSenderDelegate.ResetCounter();
+    app::CommandSender commandSender(&mockCommandSenderDelegate, &ctx.GetExchangeManager());
+
+    AddInvokeRequestData(apSuite, apContext, &commandSender);
+    asyncCommand = true;
+    err          = commandSender.SendCommandRequest(ctx.GetSessionBobToAlice());
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    ctx.DeliverOneMessage();
+
+    NL_TEST_ASSERT(apSuite,
+                   mockCommandSenderDelegate.onResponseCalledTimes == 0 && mockCommandSenderDelegate.onFinalCalledTimes == 0 &&
+                       mockCommandSenderDelegate.onErrorCalledTimes == 0);
+
+    NL_TEST_ASSERT(apSuite, GetNumActiveHandlerObjects() == 1);
+    NL_TEST_ASSERT(apSuite, ctx.GetExchangeManager().GetNumActiveExchanges() == 2);
+
+    System::PacketBufferHandle msgBuf;
+    WriteRequestMessage::Builder request;
+    System::PacketBufferTLVWriter writer;
+
+    chip::app::InitWriterWithSpaceReserved(writer, 0);
+    err = request.Init(&writer);
+    err = writer.Finalize(&msgBuf);
+    err = commandSender.mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::WriteRequest, std::move(msgBuf),
+                                                Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    ctx.DrainAndServiceIO();
+
+    asyncCommandHandle = nullptr;
```

Comment:
But where are we checking that the status response was sent, etc, etc?  Shouldn't we at least check that the sender got an error from the server?

Reply:
command handler is in async status, I am trying to use command sender to send out write request and trigger unknown message flow in handler, I am encountering the same issue in ReadInteraction test, would follow-up this test.

Reply:
https://github.com/project-chip/connectedhomeip/issues/20259